File: java8/src/main/java/com/shekhargulati/ninetynine_problems/_01_lists/P14.java
package com.shekhargulati.ninetynine_problems._01_lists;


import java.util.List;
import java.util.stream.Stream;

import static java.util.stream.Collectors.toList;

/**
 * (*) Duplicate the elements of a list
 * <pre>
 *          duplicate(Arrays.asList("a", "b", "c", "d"))
 * </pre>
 */
public class P14 {

    public static <T> List<T> duplicate(List<T> list) {
        return list.stream().flatMap(e -> Stream.of(e, e)).collect(toList());
    }
}

File: scala/src/main/scala/com/shekhargulati/ninetynine_problems/_01_lists/P14.scala
package com.shekhargulati.ninetynine_problems._01_lists

/**
  * <b>(*) Duplicate the elements of a list</b>
  *
  * {{{
  *     duplicate(List("a", "b", "c", "d"))
  * }}}
  */
object P14 {

  /**
    * Using flatMap
    */
  def duplicate[T](list: List[T]): List[T] = list.flatMap(e => List(e, e))


  /**
    * Using recursion
    */
  def duplicate_recursion[T](list: List[T]): List[T] = list match {
    case x :: xs => x +: x +: duplicate_recursion(xs)
    case Nil => Nil
  }
}
